guidance/hybridpath_guidance/hybridpath_guidance/hybridpath.py

```diff
@@ -0,0 +1,243 @@
+
+    Output data:
+    Path: A path structure with the following fields:
+        NumSubpaths: Number of subpaths
+        Order: Order of the subpaths
+        WP: Waypoints
+        LinSys: Linear system to solve the subpaths
+            A: Coefficients matrix
+            bx: x-vector for each subpath
+            by: y-vector for each subpath
+        coeff: Coefficients
+            a: x-coefficients
+            b: y-coefficients
+            a_der: x-coefficients for derivatives
+            b_der: y-coefficients for derivatives
+
+        Based on MATLAB code by prof. Roger Skjetne, NTNU
+    '''
+
+    def __init__(self, WP, r, lambda_val, plot_handle = None):
+        self.WP = WP
+        self.r = r
+        self.lambda_val = lambda_val
+        self.plot_handle = plot_handle
+        self.ord = 2*r + 1
+        self.Path = {
+            'NumSubpaths': len(WP) - 1,
+            'Order': self.ord,
+            'WP': {'x': WP[:,0], 'y': WP[:,1]},
+            'LinSys': {
+                'A': None,
+                'bx': [],
+                'by': []
+            },
+            'coeff': {
+                'a': [],
+                'b': [],
+                'a_der': [],
+                'b_der': []
+            }
+        }
+        self._calculate_subpaths()
+
+    def _calculate_subpaths(self):
+        ord_plus_one = self.ord + 1
+        A = np.zeros((ord_plus_one, ord_plus_one))
+        coefs = np.ones(ord_plus_one)
+        A[0,0] = coefs[0]
+        A[1,:] = coefs
+        c_der = coefs.copy() #
+        for k in range(2, (self.ord + 1) // 2 + 1):
+            c_der = np.polyder(c_der[::-1])[::-1]
+            coefs = np.concatenate([np.zeros(k-1), c_der])
+            #coefs = np.pad(c_der, (k-1, 0), 'constant')
+            A[2*k-2, k-1] = c_der[0]
+            A[2*k-1, :] = coefs
+        self.Path['LinSys']['A'] = A
+
+        N = self.Path['NumSubpaths'] #
+        for j in range(N):
+            ax, bx = np.zeros(ord_plus_one), np.zeros(ord_plus_one)
+            ax[:2] = self.WP[j:j+2, 0]
+            bx[:2] = self.WP[j:j+2, 1]
+            if self.ord > 2:
+                if j == 0:
+                    ax[2:4] = [self.WP[j+1, 0] - self.WP[j, 0], self.lambda_val * (self.WP[j+2, 0] - self.WP[j, 0])] 
+                    bx[2:4] = [self.WP[j+1, 1] - self.WP[j, 1], self.lambda_val * (self.WP[j+2, 1] - self.WP[j, 1])] 
+                elif j == N - 1:
+                    ax[2:4] = [self.lambda_val * (self.WP[j+1, 0] - self.WP[j-1, 0]), self.WP[j+1, 0] - self.WP[j, 0]]
+                    bx[2:4] = [self.lambda_val * (self.WP[j+1, 1] - self.WP[j-1, 1]), self.WP[j+1, 1] - self.WP[j, 1]]
+
+                else:
+                    ax[2:4] = [self.lambda_val * (self.WP[j+1, 0] - self.WP[j-1, 0]), self.lambda_val * (self.WP[j+2, 0] - self.WP[j, 0])]
+                    bx[2:4] = [self.lambda_val * (self.WP[j+1, 1] - self.WP[j-1, 1]), self.lambda_val * (self.WP[j+2, 1] - self.WP[j, 1])]
```

Comment:
I think this deserves its own method, since there is no way to tell what this actually does without reading through the chaos of indexing 😅